### PR TITLE
[DEVHAS-21] Add webhook scaffold & logic for update

### PR DESCRIPTION
**Pull request by maysunfaisal**
Signed-off-by: Maysun J Faisal <maysunaneek@gmail.com>

Fixes https://issues.redhat.com/browse/DEVHAS-21 and https://issues.redhat.com/browse/DEVHAS-20

## Summary

This PR adds the webhook validation for `Application` & `Component` update. The validating webhook blocks the client from submitting CR updates to fields that are immutable acc to [doc](https://docs.google.com/document/d/1dfkTky3qFv6wTXdMX9Ag-bYFEt3OfBgRZOceYNSwyEU/edit?pli=1#heading=h.6b6ftdufj5y3).

It also add create validation checks for the `Component` CR, to make sure at least one source either git or image is specified and if a git source is specified, the URL is a valid URL.

The `Application` spec has been updated to make sure the `displayName` is required, the `Component` spec has been updated to add regex and char limit for `application` & `componentName`

The Makefile also has new commands to install and uninstall certificate manager kube resources to add the CR for self-signed certificates as required by [admission webhooks](https://sdk.operatorframework.io/docs/building-operators/golang/webhook/) since they intercept the API request.

The PR also adds unit and integration tests. The webhook integration tests  does not check reconciler logic as the k8sClient generated by the scaffold does not add the Reconcilers to the manager for the testEnv.

## How to test it?

Create an `Application` & `Component` as you would normally do. Go either to the `Application` or the `Component` yaml that was used and update any immutable field and apply. You should see that the server blocks the update.

For example, I updated the component name in the `Component` yaml from `backend` to `backends`:
```
$ k apply -f config/samples/component/component-basic.yaml 
Error from server (component name cannot be updated to backends): error when applying patch:
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"appstudio.redhat.com/v1alpha1\",\"kind\":\"Component\",\"metadata\":{\"annotations\":{},\"name\":\"component-sample\",\"namespace\":\"application-service-system\"},\"spec\":{\"application\":\"application-sample\",\"componentName\":\"backends\",\"env\":[{\"name\":\"FOO\",\"value\":\"foo1\"},{\"name\":\"BAR\",\"value\":\"bar1\"}],\"replicas\":1,\"resources\":{\"limits\":{\"cpu\":\"2\",\"ephemeral-storage\":\"500Mi\",\"memory\":\"500Mi\",\"storage\":\"400Mi\"},\"requests\":{\"cpu\":\"700m\",\"ephemeral-storage\":\"400Mi\",\"memory\":\"400Mi\",\"storage\":\"200Mi\"}},\"route\":\"route111\",\"source\":{\"git\":{\"url\":\"https://github.com/devfile-samples/devfile-sample-java-springboot-basic\"}},\"targetPort\":1111}}\n"}},"spec":{"componentName":"backends"}}
to:
Resource: "appstudio.redhat.com/v1alpha1, Resource=components", GroupVersionKind: "appstudio.redhat.com/v1alpha1, Kind=Component"
Name: "component-sample", Namespace: "application-service-system"
for: "config/samples/component/component-basic.yaml": admission webhook "vcomponent.kb.io" denied the request: component name cannot be updated to backends
```

## Codecov

I added a new codecov config file to ignore the generated deepcopy file from the operator-sdk. Since the `v1alpha1` dir now has *_test.go files, this dir is now included in the codecov report. This PR so far increases the total coverage of the project by 3%

Screen Shot 2022-01-07 at 3 09 53 PM

### CODE REVIEW

**Comment by sbose78** [comment]
Curious, what does the webhook validate?

**Comment by maysunfaisal** [comment]
> Curious, what does the webhook validate?

@sbose78 this is to check the CR updates are being done properly. That is the client cannot update the immutable fields as outlined in [doc](https://docs.google.com/document/d/1dfkTky3qFv6wTXdMX9Ag-bYFEt3OfBgRZOceYNSwyEU/edit?pli=1#heading=h.apl9rlcsyxg8)

**Review by johnmcollier — CHANGES_REQUESTED** [review]
Really nicely done 👍 . Just one nitpicking comment about logs 

Any luck getting this running on KCP? Is installing without webhooks on KCP possible? We should reach out to them anyways, since we'll need to run on KCP at some point anyways.

File: api/v1alpha1/application_webhook.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	// TODO(user): fill in your defaulting logic.
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-application,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=applications,verbs=create;update,versions=v1alpha1,name=vapplication.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Validator = &Application{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Application) ValidateCreate() error {
```
Comment (johnmcollier) [inline api/v1alpha1/application_webhook.go:57]: We should validate that the Application display name is set, (it's required according to the spec)

File: api/v1alpha1/component_webhook.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	// TODO(user): fill in your defaulting logic.
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-component,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=components,verbs=create;update,versions=v1alpha1,name=vcomponent.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Validator = &Component{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Component) ValidateCreate() error {
```
Comment (johnmcollier) [inline api/v1alpha1/component_webhook.go:57]: Should validate the following on create:

`spec.componentName` --> Non-empty and matches the following regex: `^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
`spec.application` --> Should be non-empty

One of the following should be set:
`spec.source.git.url`
`spec.source.image.containerImage`

If `spec.source.git.url` is set, it should be a valid URL

**Review by johnmcollier — COMMENTED** [review]
Added a couple more notes on what I think should be validated upon create, otherwise have no other concerns.

File: api/v1alpha1/application_webhook.go (hunk 1 above)
Comment (maysunfaisal) [reply api/v1alpha1/application_webhook.go:56]: updated the spec

File: api/v1alpha1/component_webhook.go (hunk 2 above)
Comment (maysunfaisal) [reply api/v1alpha1/component_webhook.go:57]: updated the spec and the create webhook

File: api/v1alpha1/application_webhook.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+//+kubebuilder:webhook:path=/mutate-appstudio-redhat-com-v1alpha1-application,mutating=true,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=applications,verbs=create;update,versions=v1alpha1,name=mapplication.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Defaulter = &Application{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *Application) Default() {
+	applicationlog.Info("default", "name", r.Name)
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply api/v1alpha1/application_webhook.go]: I kept it for webhook funcs being used and called, removed from the unused scaffolds

**Comment by maysunfaisal** [comment]
@johnmcollier I didnt add a regex for application.spec.displayName, I feel it will be too restrictive on what you can and cannot put, if we also consider that it will probably be used by HAC on the UI.. I think the Sanitization is good enough if we want to use display name for our CRs 🤔

**Review by johnmcollier — APPROVED** [review]

**Pull request merged**